File: ilamicbook/backend-fastapi/src/main/python/com/ilamicbook/backend/controllers/book_controller.py
from fastapi import APIRouter, HTTPException
from typing import List
from ..models.book import Book
from ..services.book_service import BookService

router = APIRouter(prefix="/api", tags=["books"])
book_service = BookService()

@router.get("/books", response_model=List[Book])
async def get_books():
    """Get all books"""
    return book_service.get_all_books()

@router.get("/books/{book_id}", response_model=Book)
async def get_book(book_id: int):
    """Get book by ID"""
    book = book_service.get_book_by_id(book_id)
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")
    return book 

File: ilamicbook/backend-go/src/main/go/com/ilamicbook/backend/controllers/book_controller.go
package controllers

import (
	"net/http"
	"strconv"

	"com/ilamicbook/backend/services"

	"github.com/gin-gonic/gin"
)

// BookController handles HTTP requests for books
type BookController struct {
	bookService *services.BookService
}

// NewBookController creates a new BookController instance
func NewBookController(bookService *services.BookService) *BookController {
	return &BookController{
		bookService: bookService,
	}
}

// GetBooks returns all books
func (c *BookController) GetBooks(ctx *gin.Context) {
	books := c.bookService.GetAllBooks()
	ctx.JSON(http.StatusOK, books)
}

// GetBook returns a book by ID
func (c *BookController) GetBook(ctx *gin.Context) {
	idStr := ctx.Param("id")
	id, err := strconv.Atoi(idStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"error": "Invalid book ID"})
		return
	}

	book := c.bookService.GetBookByID(id)
	if book == nil {
		ctx.JSON(http.StatusNotFound, gin.H{"error": "Book not found"})
		return
	}

	ctx.JSON(http.StatusOK, book)
}
